src/sentry/db/models/query.py

```diff
@@ -80,6 +80,12 @@ def create_or_update(model, using=None, **kwargs):
         return affected, False
 
     create_kwargs = kwargs.copy()
+
+    # only use join filters for filtering (above) and not creation
+    for key in create_kwargs.keys():
+        if '__' in key:
+            create_kwargs.pop(key)
```

Comment:
This is the weirdest part of this PR.

The issue is that tag counters increments (`times_seen` and `values_seen`) start at the "bottom" and trickle "up.":

* When an event is created, we first incr the values `times_seen`, because it needs to be incremented every time.
* A signal on that save fires, which then checks whether the value as just *created* or not.
* If the value was created, we need to increment the key's `values_seen`.

All of this passes along via the buffer abstraction and signal firing. The last thing to be incremented is `TagKey`, which needs to know what `environment_id` to use. Previously, all models had an `environment_id` so it was already being passed along. Now they don't, so I'm sending an ORM join filter (`_key__environment_id`) down the chain which is handy for verifying we aren't doing something wrong and mandatory for the last guy in the chain (`TagKey`). The weird bit is we don't want these join filters to be passed to `model.objects.create`...

Reply:
If that doesn't hurt your head, here's a flow (v2 is basically the same): https://github.com/getsentry/sentry/blob/345ff81ecd853bbf9d2ff9f3cba37f6bd08350ea/src/sentry/tagstore/legacy/backend.py#L100

Reply:
Making a note to self to come back to this, I feel like there must be a clearer approach?

Reply:
The only one I could think of that doesn't entirely change the flow is adding another kwarg to buffer that is like a "pass-through." But that felt worse to me.

So then I was left with completely changing how these get incremented and written. I don't think I could use the buffer API at all in that case (if we're against the `__` check or a pass-through param) and that felt like too much to take on for little gain.

But yeah, I have been thinking about this offline as well. Open to idearz.